File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.java
import java.util.Scanner;

public class Main { // DecimalHexadecimal
    static final int LimiteMaximoDecimal = 2147483647;
    static final String LimiteMaximoHexadecimal = "7FFFFFFF";

    public static String DecimalParaHexadecimal(int decimal) {
        String hexadecimal = "";
        do {
            int resto = decimal % 16;
            if (resto >= 0 && resto <= 9)
                hexadecimal = (char) (resto + '0') + hexadecimal;
            else
                hexadecimal = (char) (resto - 10 + 'A') + hexadecimal;
            decimal = decimal / 16;
        } while (decimal != 0);
        return hexadecimal;
    }

    public static int HexadecimalParaDecimal(String hexadecimal) {
        int potencia = 0;
        int decimal = 0;
        while (!hexadecimal.isEmpty()) {
            char digito = Character.toUpperCase(hexadecimal.charAt(hexadecimal.length() - 1));
            if (digito >= '0' && digito <= '9')
                decimal += (digito - '0') * Math.pow(16, potencia);
            else if (digito >= 'A' && digito <= 'F')
                decimal += (digito - 'A' + 10) * Math.pow(16, potencia);
            hexadecimal = hexadecimal.substring(0, hexadecimal.length() - 1);
            potencia++;
        }
        return decimal;
    }

    public static boolean IsValidHexadecimal(String hexadecimal) {
        for (int i = 0; i < hexadecimal.length(); i++) {
            char digito = Character.toUpperCase(hexadecimal.charAt(i));
            if (!(digito >= '0' && digito <= '9' || digito >= 'A' && digito <= 'F'))
                return false;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int opcao;
        int decimal;
        String hexadecimal;
        do {
            System.out.println("===============================");
            System.out.println(" CONVERSOR DECIMAL-HEXADECIMAL ");
            System.out.println("===============================");
            System.out.println();
            System.out.println("1 - Converter decimal para hexadecimal");
            System.out.println("2 - Converter hexadecimal para decimal");
            System.out.println("0 - Sair");
            System.out.println();
            System.out.print("Digite a opcao desejada: ");
            opcao = scanner.nextInt();
            switch (opcao) {
                case 1:
                    System.out.println();
                    System.out.print("Digite um numero decimal (ate " + LimiteMaximoDecimal + "): ");
                    decimal = scanner.nextInt();
                    if (decimal > LimiteMaximoDecimal)
                        System.out.println("Valor decimal fora do limite permitido.");
                    else
                        System.out.println("Valor em hexadecimal: " + DecimalParaHexadecimal(decimal));
                    break;
                case 2:
                    System.out.println();
                    System.out.print("Digite um numero hexadecimal (ate " + LimiteMaximoHexadecimal + "): ");
                    hexadecimal = scanner.next();
                    if (!IsValidHexadecimal(hexadecimal))
                        System.out.println("Valor hexadecimal invalido.");
                    else if (HexadecimalParaDecimal(hexadecimal) > LimiteMaximoDecimal)
                        System.out.println("Valor hexadecimal fora do limite permitido.");
                    else
                        System.out.println("Valor em decimal: " + HexadecimalParaDecimal(hexadecimal));
                    break;
            }
            System.out.println();
        } while (opcao != 0);
    }
}

File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.dart
import 'dart:io';
import 'dart:math';

String DecimalParaHexadecimal(int decimal) {
  String hexadecimal = '';
  do {
    int resto = decimal % 16;
    if (resto >= 0 && resto <= 9)
      hexadecimal = String.fromCharCode(resto + '0'.codeUnitAt(0)) + hexadecimal;
    else
      hexadecimal =
          String.fromCharCode(resto - 10 + 'A'.codeUnitAt(0)) + hexadecimal;
    decimal = decimal ~/ 16;
  } while (decimal != 0);
  return hexadecimal;
}

int HexadecimalParaDecimal(String hexadecimal) {
  int potencia = 0;
  int decimal = 0;
  while (hexadecimal.isNotEmpty) {
    String digito = hexadecimal[hexadecimal.length - 1].toUpperCase();
    if (digito.compareTo('0') >= 0 && digito.compareTo('9') <= 0)
      decimal +=
          int.parse(digito) * pow(16, potencia).toInt();
    else if (digito.compareTo('A') >= 0 && digito.compareTo('F') <= 0)
      decimal += (digito.codeUnitAt(0) - 'A'.codeUnitAt(0) + 10) *
          pow(16, potencia).toInt();
    hexadecimal = hexadecimal.substring(0, hexadecimal.length - 1);
    potencia++;
  }
  return decimal;
}

bool IsValidHexadecimal(String hexadecimal) {
  RegExp regex = RegExp(r'^[0-9A-Fa-f]+$');
  return regex.hasMatch(hexadecimal);
}

void main() {
  const int LimiteMaximoDecimal = 2147483647;
  const String LimiteMaximoHexadecimal = '7FFFFFFF';
  int opcao;
  int decimal;
  String hexadecimal;
  do {
    print('===============================');
    print(' CONVERSOR DECIMAL-HEXADECIMAL ');
    print('===============================');
    print('');
    print('1 - Converter decimal para hexadecimal');
    print('2 - Converter hexadecimal para decimal');
    print('0 - Sair');
    print('');
    stdout.write('Digite a opcao desejada: ');
    opcao = int.parse(stdin.readLineSync()!);
    switch (opcao) {
      case 1:
        print('');
        stdout.write(
            'Digite um numero decimal (ate $LimiteMaximoDecimal): ');
        decimal = int.parse(stdin.readLineSync()!);
        if (decimal > LimiteMaximoDecimal)
          print('Valor decimal fora do limite permitido.');
        else
          print(
              'Valor em hexadecimal: ${DecimalParaHexadecimal(decimal)}');
        break;
      case 2:
        print('');
        stdout.write(
            'Digite um numero hexadecimal (ate $LimiteMaximoHexadecimal): ');
        hexadecimal = stdin.readLineSync()!;
        if (!IsValidHexadecimal(hexadecimal))
          print('Valor hexadecimal invalido.');
        else if (HexadecimalParaDecimal(hexadecimal) > LimiteMaximoDecimal)
          print('Valor hexadecimal fora do limite permitido.');
        else
          print('Valor em decimal: ${HexadecimalParaDecimal(hexadecimal)}');
        break;
    }
    print('');
  } while (opcao != 0);
}
